### ISSUE TITLE
`repeat_dim` backward folds the gradient as for an interleaved repeat, but the forward tiles

### ISSUE THREAD

**Issue by swfsql** [tracel-ai/burn#5833]
*(Analysis and reproduction prepared by Claude)*

**Describe the bug**

`Tensor::repeat_dim(dim, times)` tiles: `[a b c].repeat_dim(0, 2)` is `[a b c a b c]`, as
documented, and flex, ndarray and CUDA agree. Its autodiff backward, however, folds the output
gradient as for an **interleaved** repeat (`[a a b b c c]`, i.e. `repeat_interleave`). With
`n` the size of `dim`, the gradient of the input is

- expected: `Σ_k dy[…, k·n + i, …]`, i.e. `Σ_k dy.narrow(dim, k·n, n)`;
- actual: `Σ_k dy[…, i·times + k, …]`, and it matches that sum exactly.

There is no error and no NaN, only wrong gradients. This happens at every dim whose size is
greater than 1: leading, middle, and last. `Tensor::repeat` calls `repeat_dim` once per dim,
so it is wrong too.

The backward
([`burn-autodiff/src/ops/tensor.rs#L4644-L4656`](https://github.com/tracel-ai/burn/blob/d85aa0ab39dbf5d9a8b911081485dbfc9d0d5aa2/crates/burn-autodiff/src/ops/tensor.rs#L4644-L4656)):

```rust
unary::<B, _>(ops.parents, ops.node, grads, |grad| {
    let mut dims = grad.shape();
    let orig_dim_size = dims[dim] / times;
    if orig_dim_size > 1 {
        dims[dim] = orig_dim_size;
        let orig_dims = dims.clone();
        dims.insert(dim + 1, times); // shape [..., orig_dim_size, times, ...]
        let grad = B::float_reshape(grad, dims);
        let grad = B::float_sum_dim(grad, dim + 1); // sum over repeat times
        B::float_reshape(grad, orig_dims)
    } else {
        B::float_sum_dim(grad, dim)
    }
});
```

A row-major reshape of the repeated dim (size `times·n`) to `[n, times]` puts element
`i·times + k` at `[i, k]`. That is the layout of an interleaved repeat. The forward's
layout is `[times, n]`: element `k·n + i` is at `[k, i]`. When `n == 1`, both layouts are
the same, so the `else` branch is correct. The layout arrived with #2777, which added the
reshape for `n > 1` (before it, the backward was only `sum_dim`, which gives the wrong
shape).

The existing tests
([`burn-backend-tests/tests/autodiff/repeat_dim.rs`](https://github.com/tracel-ai/burn/blob/d85aa0ab39dbf5d9a8b911081485dbfc9d0d5aa2/crates/burn-backend-tests/tests/autodiff/repeat_dim.rs))
do not catch it. In both, the gradient that gets to the repeat is `tensor_1ᵀ · 1`, which is
constant along the repeated dim. With such a gradient, every folding gives the same sum.

**To Reproduce**

`Cargo.toml` (burn `d85aa0ab`, main 2026-09-24):

```toml
[package]
name = "repeat-dim"
version = "0.1.0"
edition = "2024"

[dependencies.burn]
git = "https://github.com/tracel-ai/burn.git"
rev = "d85aa0ab39dbf5d9a8b911081485dbfc9d0d5aa2"
default-features = false
features = ["autodiff", "flex", "std"]
```

`src/main.rs`:

```rust
use burn::prelude::*;

fn main() {
    let device = Device::flex().autodiff();
    let h = Tensor::<1>::from_floats([1.0, 2.0, 3.0], &device).require_grad();
    let w = Tensor::<1>::from_floats([1.0, 2.0, 3.0, 4.0, 5.0, 6.0], &device);

    let y = h.clone().repeat_dim(0, 2);
    println!("y     = {}", y.to_data()); // tiled: [1, 2, 3, 1, 2, 3]

    let grads = (y * w).sum().backward();
    println!("dL/dh = {}", h.grad(&grads).unwrap().to_data());
    // tiled       (right): dL/dh[i] = w[i] + w[i + 3]  = [5, 7, 9]
    // interleaved (wrong): dL/dh[i] = w[2i] + w[2i + 1] = [3, 7, 11]
}
```

Output:

```text
y     = [1.0, 2.0, 3.0, 1.0, 2.0, 3.0]
dL/dh = [3.0, 7.0, 11.0]
```

A fuller check: `h` and `w` are random, `loss = (h.repeat_dim(dim, times) · w).sum()`, and
the gradient of `h` is compared with the two foldings of `w`, both computed on the host.
The result is the same on flex and on ndarray. CUDA draws different random values, but the
pattern is the same: `dh` is 1.96 to 6.23 from the tiled sum, and ≤ 4.8e-7 from the
interleaved sum, except for the size-1 dim, where it is 0 from both.

| `h` shape | op | max \|y − tiled\| | max \|dh − tiled sum\| | max \|dh − interleaved sum\| |
|---|---|---|---|---|
| `[1, 2, 4]` | `repeat_dim(2, 3)` | 0 | 7.38 | 0 |
| `[2, 3, 4]` | `repeat_dim(0, 2)` | 0 | 2.57 | 0 |
| `[2, 3, 4]` | `repeat_dim(1, 2)` | 0 | 3.11 | 0 |
| `[2, 3, 4]` | `repeat_dim(2, 2)` | 0 | 3.40 | 0 |
| `[2, 3, 4]` | `repeat_dim(0, 3)` | 0 | 1.67 | 0 |
| `[2, 3, 4]` | `repeat_dim(1, 3)` | 0 | 4.91 | 0 |
| `[2, 3, 4]` | `repeat_dim(2, 3)` | 0 | 4.12 | 0 |
| `[2, 1, 4]` | `repeat_dim(1, 3)` | 0 | 0 | 0 |
| `[3]` | `repeat_dim(0, 2)` | 0 | 0.567 | 0 |
| `[2, 3]` | `repeat(&[2, 3])` | 0 | 5.74 | 4.8e-7 |
| `[2, 3, 4]` | `repeat(&[2, 1, 3])` | 0 | 6.12 | 2.4e-7 |

**Expected behavior**

`dL/dh = [5.0, 7.0, 9.0]`. The gradient of `repeat_dim` is the sum of the `times` tiles
of the output gradient, `Σ_k dy.narrow(dim, k·n, n)`.

**Fix**

Put `times` before `n` in the reshape, and sum that dim:

```diff
--- a/crates/burn-autodiff/src/ops/tensor.rs
+++ b/crates/burn-autodiff/src/ops/tensor.rs
@@ -4648,9 +4648,9 @@
                     if orig_dim_size > 1 {
                         dims[dim] = orig_dim_size;
                         let orig_dims = dims.clone();
-                        dims.insert(dim + 1, times); // shape [..., orig_dim_size, times, ...]
+                        dims.insert(dim, times); // shape [..., times, orig_dim_size, ...]
                         let grad = B::float_reshape(grad, dims);
-                        let grad = B::float_sum_dim(grad, dim + 1); // sum over repeat times
+                        let grad = B::float_sum_dim(grad, dim); // sum over repeat times
                         B::float_reshape(grad, orig_dims)
                     } else {
                         B::float_sum_dim(grad, dim)
```

With this change, the program above prints `dL/dh = [5.0, 7.0, 9.0]`. In every row of the
table, `dh` then matches the tiled sum (≤ 2.4e-7) on flex, on ndarray and on CUDA.

Below is a regression test for `burn-backend-tests/tests/autodiff/repeat_dim.rs`. Its
output gradient changes along the repeated dim. On flex, main fails it in both the `base`
and the `checkpointing` suites, and the values it gets are the interleaved sums:

```text
Tensors are not eq:
  => Position 0: 3 != 5
  => Position 2: 11 != 9
  => Position 3: 15 != 17
  => Position 5: 23 != 21
```

With the fix, it passes. The four existing tests pass with and without the fix.

```rust
#[test]
fn should_diff_repeat_non_uniform_grad() {
    // `repeat_dim` tiles: [a b c] -> [a b c a b c]. The output gradient must vary along the
    // repeated dim, or tiled and interleaved sums agree.
    let data_1 = TensorData::from([[1.0, 2.0, 3.0], [4.0, 5.0, 6.0]]);
    let data_2 = TensorData::from([
        [1.0, 2.0, 3.0, 4.0, 5.0, 6.0],
        [7.0, 8.0, 9.0, 10.0, 11.0, 12.0],
    ]);

    let device = AutodiffDevice::new();
    let tensor_1 = TestTensor::<2>::from_data(data_1, &device).require_grad();
    let tensor_2 = TestTensor::<2>::from_data(data_2, &device);

    let tensor_3 = tensor_1.clone().repeat_dim(1, 2).mul(tensor_2);
    let grads = tensor_3.sum().backward();

    let grad_1 = tensor_1.grad(&grads).unwrap();

    grad_1.to_data().assert_eq(
        &TensorData::from([[5.0, 7.0, 9.0], [17.0, 19.0, 21.0]]),
        false,
    );
}
```

**Additional context**

A packed-sequence test in a downstream crate (burn-mamba) found this. There, a tiling built
as `cat([head, zeros]).repeat_dim(dim, nchunks)` gave wrong gradients to `head`. A
workaround until the fix: a `gather` by the index `p mod n`. Its backward is a
scatter-add, so it is correct.

- _user: Potentially related: https://github.com/tracel-ai/burn/issues/4688_

**Environment**

- burn `d85aa0ab` (main, 2026-09-24, `0.22.0-pre.4`)
- backends flex and ndarray (CPU), and CUDA (RTX 2060, sm_75, CUDA 12.2), f32. The bug is
  in `burn-autodiff`, so it does not depend on the backend.
- rustc 1.97.1, release build
